tests/virt_autotest/sriov_network_card_pci_passthrough.pm

```diff
@@ -215,48 +218,63 @@ sub prepare_guest {
-        }
-        else {
-            assert_script_run "sed -i \"/<domain /a\\<features>\\n<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\\n</features>\" $vm.xml";
-        }
-
-        #disable memory ballooning for fv guest as it is not supported
-        if (is_fv_guest($vm)) {
-            assert_script_run "sed -i '/<currentMemory/d' $vm.xml";
+    my $mem          = "";
+    my $cur_mem      = "";
+    my $is_ballooned = 0;
+    $mem          = script_output "xmlstarlet sel -t -v //memory $vm.xml";
+    $cur_mem      = script_output "xmlstarlet sel -t -v //currentMemory $vm.xml";
+    $is_ballooned = 1 if $cur_mem != $mem;
+
+    unless (is_sle('=12-SP5') && (is_kvm_host || (is_fv_guest($vm) && !$is_ballooned))) {
+
+        script_run "virsh destroy $vm";
+        assert_script_run "virsh undefine $vm";
+
+        if (is_sle('>=15-SP2') && is_kvm_host) {
+            #for sles15sp1+, PCIe replaces PCI. We need add pcie controllers to support hotplugging more SR-IOV Ethernet vf devices
+            my $controller_xml = "<controller type='pci' model='pcie-root-port'/>";
+            assert_script_run "sed -i \"/<devices>/a\\$controller_xml\\n$controller_xml\\n$controller_xml\" $vm.xml";
         }
 
-        #enable udev debug logs
-        my $udev_conf_file = "/etc/udev/udev.conf";
-        if (script_run("ssh root\@$vm \"ls $udev_conf_file\"") == 0) {
-            script_run "ssh root\@$vm \"sed -i '/udev_log *=/{h;s/^[# ]*udev_log *=.*\$/udev_log=debug/};\${x;/^\$/{s//udev_log=debug/;H};x}' $udev_conf_file";
+        elsif (is_xen_host) {
+            #disable memory ballooning for fv guest as it is not supported
+            if (is_fv_guest($vm) && $is_ballooned) {
+                assert_script_run "sed -i '/<currentMemory/d' $vm.xml";
+                record_info "Disable guest ballooning", "$vm";
+            }
+            #enable pci-passthrough and set e820_host for pv guest
+            #enable pci-passthrough for fv guest on sles15sp2+
+            #refer to bug #1167217 for the reason
+            #but need wait libvirt support for pv guest requested in a bug on sles12sp5
+            if (is_pv_guest($vm) || (is_fv_guest($vm) && is_sle('>15-SP2'))) {
+                my $passthru_xml = "<passthrough state='on'/>";
+                my $e820_xml     = "";
+                if (is_pv_guest($vm)) {
+                    $e820_xml = "<e820_host state='on'/>";
+                }
+                if (script_run("grep '<features>' $vm.xml") == 0) {
+                    assert_script_run "sed -i \"/<features>/a\\<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\" $vm.xml";
+                }
+                else {
+                    assert_script_run "sed -i \"/<domain /a\\<features>\\n<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\\n</features>\" $vm.xml";
+                }
+            }
         }
 
+        assert_script_run "virsh define $vm.xml";
+        assert_script_run "virsh start $vm";
+        sleep 60;
```

Comment:
You can wait_still_screen(60) I think

Reply:
No, the screen change does not reflect guest fully bootup.